### PR TITLE
Simplify tracking Container loading states and address bug:

### PR DESCRIPTION
**Pull request by vladsud**
Simplify tracking Container loading states and address bug:
We may try to raise "connecting" event when context is not fully loaded, but we transitioned to closed or closing state.
See PR #10572 for discussion

Also moved "ContainerClose" event up to ensure we capture lifecycle state before it changed to "closing" state

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -1629,10 +1615,14 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         }
 
         const state = this.connectionState === ConnectionState.Connected;
+
+        assert(this.context !== undefined,
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:1619]: you will never hit this assert, if this._context is undefined, this.context will throw as it never returns undefined

**Review by anthony-murphy — APPROVED** [review]
thanks for the fix. looks good. i would love to see some refactoring is this class eventually, possibly encapsulating all interactions with the context to ease some of the complexity around the intersecting state machines. i don't have a super clear picture of what that would look like as i'm not very familiar with this code, but right now it's really easy to make mistakes given all the subtleties

**Comment by anthony-murphy** [comment]
since it looks like this has build issue, i modified my original PR which i think more closely matches you intent:
https://github.com/microsoft/FluidFramework/pull/10572/files

still work needed here, but looking to move past this error, as it creates a lot of noise

**Comment by vladsud** [comment]
I fixed it. A bit ugly, the whole shutdown sequence is a bit ugly :)

**Pull request merged**